Fix fog color being inaccurate

rafalh opened the pull request:
# Objective

Fog color was passed to shaders without conversion from sRGB to linear color space. Because shaders expect colors in linear space this resulted in wrong color being used. This is most noticeable in open scenes with dark fog color and clear color set to the same color. In such case background/clear color (which is properly processed) is going to be darker than very far objects.

Example:
![image](https://github.com/bevyengine/bevy/assets/160391/89b70d97-b2d0-4bc5-80f4-c9e8b8801c4c)
[bevy-fog-color-bug.zip](https://github.com/bevyengine/bevy/files/13063718/bevy-fog-color-bug.zip)

## Solution

Add missing conversion of fog color to linear color space.

---

## Changelog

* Fixed conversion of fog color

## Migration Guide

-  Colors in `FogSettings` struct (`color` and `directional_light_color`) are now sent to the GPU in linear space. If you were using `Color::rgb()`/`Color::rgba()` and would like to retain the previous colors, you can quickly fix it by switching to `Color::rgb_linear()`/`Color::rgba_linear()`.

superdump commented:
I think @coreh should comment on whether fog colour is supposed to be specified as non-linear sRGB or linear. If it is meant to be specified as linear then no conversion should be done.

coreh reviewed (approved):
Hey, nice catch! I think what got me confused when setting this up originally is that with `hdr: off` the clear color is not tonemapped, only the meshes, so I could never get the fog to match it, and assumed the issue was with the clear color and not the fog. But for `hdr: on` it should now match perfectly.

Before we merge this, can you update the examples, so that they don't look darker with the new (correct) color format being sent to the GPU?

It should be as simple as replacing `Color::rgb` and `Color::rgba` with `Color::rgb_linear` and `Color::rgba_linear`.

### examples/fog.rs
```rust
FogSettings {
    color: Color::rgba_linear(0.05, 0.05, 0.05, 1.0),
    falloff: FogFalloff::Linear {
        start: 5.0,
        end: 20.0,
    },
    ..default()
},
```

### examples/atmospheric_fog.rs
```rust
FogSettings {
    color: Color::rgba_linear(0.1, 0.2, 0.4, 1.0),
    directional_light_color: Color::rgba_linear(1.0, 0.95, 0.75, 0.5),
    directional_light_exponent: 30.0,
    falloff: FogFalloff::from_visibility_colors(
        15.0, // distance in world units up to which objects retain visibility (>= 5% contrast)
        Color::rgb(0.35, 0.5, 0.66), // atmospheric extinction color (after light is lost due to absorption by atmospheric particles)
        Color::rgb(0.8, 0.844, 1.0), // atmospheric inscattering color (light gained due to scattering from the sun)
    ),
},
```

(I double checked an in this second one, we _don't_ need to change the values in `FogFalloff` since it was already correctly using linear space internally to calculate `extinction` and `inscattering`.

I believe we also need a migration guide for this. Something like: “Fog `color` and `directional_light_color` are now sent to the GPU in linear space. If you were using `Color::rgb()`/`Color::rgba()` and would like to retain the previous colors, you can quickly fix it by switching to `Color::rgb_linear()`/`Color::rgba_linear()`.”

alice-i-cecile commented:
Great work, thank you very much :) And thanks @coreh for the high quality review!

rafalh commented:
I adjusted configuration in examples to match the previous look and added migration info. I agree it may be helpful.
I didn't replace `Color::rgba` calls to `Color::rgba_linear` because I believe we shouldn't encourage people to use linear space outside of shaders. It is less intuitive to set up and for me it is a technical detail that shaders do calculations in linear space. Let me know if you disagree and I can change examples to linear values.
